perceval/backends/github.py

```diff
@@ -0,0 +1,250 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2015 Bitergia
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+#
+# Authors:
+#   Alvaro del Castillo San Felix <acs@bitergia.com>
+#
+
+
+
+import json
+import logging
+import os.path
+
+import requests
+
+from ..backend import Backend, BackendCommand
+from ..cache import Cache
+from ..errors import CacheError
+from ..utils import DEFAULT_DATETIME, str_to_datetime
+
+logger = logging.getLogger(__name__)
+
+class GitHub(Backend):
+    """GitHub backend for Perceval
+
+    This class allows the fetch the issues stored in GitHub
+    repository.
+    """
+
+    def __init__(self, owner=None, repository=None, token=None,
+                 cache=None):
+        """ Init a GitHub instance
+        :param owner: GitHub owener
+        :param repository: GitHub repository from the owner
+        :param token: GitHub auth token to access the API
+        :param cache: Use issues already retrieved in cache
+        """
+        super().__init__(cache=cache)
+        self.owner = owner
+        self.repository = repository
+        self.client = GitHubClient(owner, repository, token)
+
+    def fetch(self, from_date=DEFAULT_DATETIME):
+        """Fetch the issues from the repository.
+
+        The method retrieves, from a GitHub repository, the issues
+        updated since the given date.
+
+        :param from_date: obtain issues updated since this date
+
+        :returns: a generator of issues
+        """
+
+        self._purge_cache_queue()
+
+        issues = self.client.get_issues(from_date)
+
+        self._push_cache_queue(issues)
+        self._flush_cache_queue()
+
+        while issues:
+            issue = issues.pop(0)
+            yield issue
+
+            if not issues:
+                issues = self.client.get_issues()
```

Comment:
Yes, probably it could implemented as an iterator and also, as a generator. We can talk later about it, in order to be sure we add value doing that and reworking the logic.